app/index.tsx

```diff
@@ -153,54 +141,54 @@ export default class Root extends React.Component<{}, IState> {
 		store.dispatch(appInit());
 	};
 
-	getMasterDetail = (width: number) => {
+	const getMasterDetail = (width: number) => {
 		if (!isTablet) {
 			return false;
 		}
 		return width > MIN_WIDTH_MASTER_DETAIL_LAYOUT;
 	};
 
-	setMasterDetail = (width: number) => {
-		const isMasterDetail = this.getMasterDetail(width);
+	const setMasterDetail = (width: number) => {
+		const isMasterDetail = getMasterDetail(width);
 		store.dispatch(setMasterDetailAction(isMasterDetail));
 	};
 
 	// Dimensions update fires twice
-	onDimensionsChange = debounce(({ window: { width, height, scale, fontScale } }: { window: IDimensions }) => {
-		this.setDimensions({
+	const onDimensionsChange = debounce(({ window: { width, height, scale, fontScale } }: { window: IDimensions }) => {
+		setDimensions({
 			width,
 			height,
 			scale,
 			fontScale
 		});
-		this.setMasterDetail(width);
+		setMasterDetail(width);
 	});
 
-	setTheme = (newTheme = {}) => {
-		// change theme state
-		this.setState(
-			prevState => newThemeState(prevState, newTheme as IThemePreference),
-			() => {
-				const { themePreferences } = this.state;
-				// subscribe to Appearance changes
-				subscribeTheme(themePreferences, this.setTheme);
-			}
+	const setThemeFunction = (newThemeObject = {}) => {
+		const { theme: newTheme, themePreferences: newThemePreferences } = newThemeState(
+			themePreferences,
+			newThemeObject as IThemePreference
 		);
+		setThemePreferences(newThemePreferences);
+		setTheme(newTheme);
+		subscribeTheme(themePreferences, setThemeFunction);
 	};
 
-	setDimensions = ({ width, height, scale, fontScale }: IDimensions) => {
-		this.setState({ width, height, scale, fontScale });
+	const setDimensions = ({ width, height, scale, fontScale }: IDimensions) => {
+		setWidth(width);
+		setHeight(height);
+		setScale(scale);
+		setFontScale(fontScale);
```

Comment:
We'll need to test this on #4316, which uses React 18, because if automatic batching isn't working on React + RN, this is going to re-render this component 4 times.
https://reactjs.org/blog/2022/03/29/react-v18.html#new-feature-automatic-batching

Reply:
Here, the React also batch the setState, instead of re-render 4 times. However, one interesting thing that I noticed just after you commented here is that it re-renders 2 times when changing the orientation, because the orientation active the debounce, and debounce is an asynchronous function when we call setState from synchronous function and the asynchronous function will re-render multiple times. Because of that, I needed to use the [unstable_batchedUpdates](https://stackoverflow.com/a/48610973) from react-native, but I need to test when change to React 18